### PR TITLE
fix the reindexing gap

### PR DESCRIPTION
**Pull request by czue**
this does two main things:
1. ties elastic pillow checkpoint IDs to the elastic indices (and migrates the current checkpoints)
2. sets a checkpoint on reindex for the first pillow that touches the index

i'm not totally happy with this. in particular it feels pretty hacky that, e.g. the 'form' reindexer sets the checkpoint and the 'form-sql' doesn't (i think this is currently necessary due to the fact that they will be run in serial and therefore by the time 'form-sql' runs it would be setting a checkpoint that is now too late and would cause us to miss some couch forms that have come in while the 'form' reindexer was running). I think the better thing to do would have the notion of a "reindexer-group" or something, but I haven't done that yet.

i also want to clean up how we're using pillow_id a little, and do a bit more testing.

@snopoke @emord cc @benrudolph

### CODE REVIEW

File: corehq/pillows/group.py [hunk 1]
```diff
@@ -4,20 +4,19 @@
 from corehq.elastic import get_es_new
 
 from .mappings.group_mapping import GROUP_INDEX_INFO
-from pillowtop.checkpoints.manager import PillowCheckpoint, PillowCheckpointEventHandler
+from pillowtop.checkpoints.manager import PillowCheckpoint, PillowCheckpointEventHandler, \
```
Comment (dimagimon) [inline corehq/pillows/group.py]: F401 'PillowCheckpoint' imported but unused

File: corehq/pillows/reportxform.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 from corehq.pillows.base import convert_property_dict
 from corehq.pillows.mappings.reportxform_mapping import REPORT_XFORM_INDEX_INFO
 from corehq.pillows.xform import transform_xform_for_elasticsearch, xform_pillow_filter
-from pillowtop.checkpoints.manager import PillowCheckpoint
+from pillowtop.checkpoints.manager import PillowCheckpoint, get_checkpoint_for_elasticsearch_pillow
```
Comment (dimagimon) [inline corehq/pillows/reportxform.py]: F401 'PillowCheckpoint' imported but unused

File: testapps/test_pillowtop/tests/test_reindexer.py [hunk 3]
```diff
@@ -23,8 +23,11 @@
 from corehq.pillows.mappings.user_mapping import USER_INDEX
 from corehq.pillows.mappings.xform_mapping import XFORM_INDEX
 from corehq.util.elastic import delete_es_index, ensure_index_deleted
-from corehq.util.test_utils import trap_extra_setup, create_and_save_a_form, create_and_save_a_case
+from corehq.util.test_utils import trap_extra_setup, create_and_save_a_form, create_and_save_a_case, generate_cases
+from pillowtop.checkpoints.manager import DEFAULT_EMPTY_CHECKPOINT_SEQUENCE
```
Comment (dimagimon) [inline testapps/test_pillowtop/tests/test_reindexer.py]: F401 'DEFAULT_EMPTY_CHECKPOINT_SEQUENCE' imported but unused

File: corehq/apps/cleanup/migrations/0012_add_es_index_to_checkpoint_ids.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            assert new_id in current_checkoint_ids
+        except DjangoPillowCheckpoint.DoesNotExist:
+            print 'warning: legacy pillow checkpoint with ID {} not found'.format(old_id)
+        except AssertionError:
+            print 'warning: no current pillow found with checkpoint ID {}'.format(new_id)
+
+
+def delete_checkpoints(apps, schema_editor):
+    DjangoPillowCheckpoint = apps.get_model('pillowtop', 'DjangoPillowCheckpoint')
+    for _, new_id in RENAMES:
```
Comment (emord) [inline corehq/apps/cleanup/migrations/0012_add_es_index_to_checkpoint_ids.py]: I don't understand this for loop. neither variable is used

File: corehq/apps/cleanup/migrations/0012_add_es_index_to_checkpoint_ids.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            checkpoint.checkpoint_id = new_id
+            checkpoint.save()
+            assert new_id in current_checkoint_ids
+        except DjangoPillowCheckpoint.DoesNotExist:
+            print 'warning: legacy pillow checkpoint with ID {} not found'.format(old_id)
+        except AssertionError:
+            print 'warning: no current pillow found with checkpoint ID {}'.format(new_id)
+
+
+def delete_checkpoints(apps, schema_editor):
```
Comment (emord) [inline corehq/apps/cleanup/migrations/0012_add_es_index_to_checkpoint_ids.py]: should probably be `delete_default_pillow_checkpoint` ?

File: corehq/ex-submodules/pillowtop/reindexer/reindexer.py [hunk 6]
```diff
@@ -82,6 +81,12 @@ def _prepare_index_for_usage(es, index_info):
     es.indices.refresh(index_info.index)
 
 
+def _set_checkpoint_id(pillow):
```
Comment (snopoke) [inline corehq/ex-submodules/pillowtop/reindexer/reindexer.py]: rename: `_set_checkpoint`

**Comment by snopoke** [comment]
For pillows with multiple reindexers you could make it consistent across reindexers by only setting the checkpoint value for the topic being reindexed.

**Comment by czue** [comment]
> For pillows with multiple reindexers you could make it consistent across reindexers by only setting the checkpoint value for the topic being reindexed.

this is a good idea though i might punt it depending how hard it ends up looking.

**Comment by czue** [comment]
>> For pillows with multiple reindexers you could make it consistent across reindexers by only setting the checkpoint value for the topic being reindexed.

> this is a good idea though i might punt it depending how hard it ends up looking.

I think I am going to punt this. there's not a clean way to bring the notion of topics into the checkpoints without breaking a lot of abstraction  barriers or changing quite a fair amount of code.

I also kind of punted the pillow ID thing for now, though I think maybe we should just get rid of that argument in all of the pillow generator functions and hard code it. anyway I think this should be an acceptable minimum change once build passes.

File: corehq/apps/cleanup/migrations/0012_add_es_index_to_checkpoint_ids.py (hunk 4 above)
Comment (emord) [inline corehq/apps/cleanup/migrations/0012_add_es_index_to_checkpoint_ids.py]: I think this was merged without being updated. Doubt that the reverse migration is a big deal though

Comment (czue) [inline corehq/apps/cleanup/migrations/0012_add_es_index_to_checkpoint_ids.py]: it's updated below

Comment (emord) [inline corehq/apps/cleanup/migrations/0012_add_es_index_to_checkpoint_ids.py]: Ah. I see now :+1:

**Pull request merged**